src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -1,243 +1,249 @@
-  let s = name s in
-  A.pvar s
+(** [env] contains the list of recursive types during the derivation
+
+    i.e:
+    - contains one element maximum if its a single type declaration *)
+type env = string list ref
+
+(** [curr_types] saves both the current type and all mutual
+    recursive types *)
+let curr_types : (string * string list) option ref = ref None
+
+let set_curr_types x xs = curr_types := Some (x, xs)
+
+let clean_curr_types () = curr_types := None
 
-let gen ~loc ?(env = TypeGen.empty) lg =
+let curr_type () = Option.get !curr_types |> fst
+
+let curr_types () = Option.get !curr_types |> snd
+
+let env : env = ref []
+
+let clean_env () = env := []
+
+let is_rec x = List.mem x !env
+
+let add_env typ_names typ_name ty =
+  if is_rec_type_decl typ_names ty then
+    env := typ_name :: !env
+
+
+(** {2. Generator constructors} *)
+
+(** [gen lg args] creates a generator using [lg].
+
+    The longident can either be a:
+    - Lident s: We transform to gen_s (or gen if s = "t")
+    - Ldot (lg, s): We transform to qualified generator (e.g. B.gen)
+*)
+let gen ~loc lg args =
   let (module A) = Ast_builder.make loc in
   match lg with
   | Lident s ->
-      Option.value ~default:(name s |> A.evar) @@ TypeGen.find_opt s env
-  | Ldot (lg, s) -> A.(pexp_ident (Located.mk @@ Ldot (lg, name s)))
+     if is_rec s then
+       name ~sized:true s |> A.evar |>
+         Args.apply_args ~loc args |>
+         Args.apply_args ~loc [ [%expr (n / 2)] ]
+     else
+       name s |> A.evar |> Args.apply_args ~loc args
+  | Ldot (lg, s) ->
+     A.(pexp_ident (Located.mk @@ Ldot (lg, name s))) |>
+       Args.apply_args ~loc args
   | Lapply (_, _) -> raise (Invalid_argument "gen received an Lapply")
 
-let frequency ~loc l = [%expr QCheck.Gen.frequency [%e l]]
+(** [tree nodes leaves] creates a tree like generator *)
+let tree ~loc nodes leaves =
+    [%expr
+        match n with
+        | 0 -> [%e leaves]
+        | n -> [%e nodes]
+    ]
 
-let pure ~loc x = [%expr QCheck.Gen.pure [%e x]]
+(** [sized typ_name is_rec to_gen xs] uses [is_rec] to determine recursive
+    nodes in [xs].
 
-let tree ~loc nodes leaves =
-  [%expr
-    QCheck.Gen.sized
-    @@ QCheck.Gen.fix (fun self -> function
-         | 0 -> [%e leaves] | n -> [%e nodes])]
+    If no recursive node is found, the type is _not_ recursive, we build a
+    generator using frequency.
 
-let sized ~loc ~env typ_name (is_rec : 'a -> bool)
-    (to_gen : ?env:expression TypeGen.t -> 'a -> expression) (xs : 'a list) =
+    However, if recursive nodes are found, we build a tree like generator using
+    {!tree}.
+
+    The function is generalized for variants and polymorphic variants:
+
+    {[
+    type t = Leaf | Node of int * t * t
+
+    (* or *)
+
+    type t = [`Leaf | `Node of int * t * t]
+    ]}
+
+    Therefore, [is_rec] and [to_gen] are different for variants and polymorphic
+    variants. *)
+let sized ~loc (is_rec : 'a -> bool) (to_gen : 'a -> expression) (xs : 'a list) =
```

Comment:
Rename to `gen_sized`?

Reply:
Done in d527aa353bf3f971791ff1a3e112d8d2d24179d1